src/boot/efi/boot.c

```diff
@@ -1871,6 +1872,115 @@ static bool is_sd_boot(EFI_FILE *root_dir, const char16_t *loader_path) {
         return memcmp(content, magic, sizeof(magic)) == 0;
 }
 
+static EFI_STATUS launch_internal_uefi_shell(const ConfigEntry *entry) {
+        EFI_STATUS err;
+
+        VENDOR_DEVICE_PATH *dp;
+        err = BS->HandleProtocol(entry->device, MAKE_GUID_PTR(EFI_DEVICE_PATH_PROTOCOL), (void **) &dp);
+        if (err != EFI_SUCCESS)
+                return err;
+
+        if (dp->Header.Type != MEDIA_DEVICE_PATH || dp->Header.SubType != MEDIA_PIWG_FW_VOL_DP)
+                return EFI_INVALID_PARAMETER;
+
+        static struct {
+                VENDOR_DEVICE_PATH fv;
+                VENDOR_DEVICE_PATH ff;
+                EFI_DEVICE_PATH end;
+        } _packed_ internal_uefi_shell_dp = {
+                .fv = {
+                        .Header = {
+                                .Type = MEDIA_FILEPATH_DP,
+                                .SubType = MEDIA_PIWG_FW_VOL_DP,
+                                .Length = sizeof(internal_uefi_shell_dp.fv),
+                        },
+                },
+                .ff = {
+                        .Header = {
+                                .Type = MEDIA_FILEPATH_DP,
+                                .SubType = MEDIA_PIWG_FW_FILE_DP,
+                                .Length = sizeof(internal_uefi_shell_dp.ff),
+                        },
+                        .Guid = UEFI_SHELL_GUID,
+                },
+                .end = DEVICE_PATH_END_NODE,
+        };
+
+        internal_uefi_shell_dp.fv.Guid = dp->Guid;
+
+        _cleanup_(unload_imagep) EFI_HANDLE shell;
+        err = BS->LoadImage(false, IMG, (EFI_DEVICE_PATH *) &internal_uefi_shell_dp, NULL, 0, &shell);
+        if (err != EFI_SUCCESS)
+                return log_error_status(err, "Error loading internal UEFI shell: %m");
+
+        err = BS->StartImage(shell, NULL, NULL);
+        if (err != EFI_SUCCESS)
+                return log_error_status(err, "Error starting internal UEFI shell: %m");
+
+        return EFI_SUCCESS;
+}
+
+static void config_entry_add_loader_uefi_internal(Config *config) {
+        EFI_STATUS err;
+
+        /* Finds the internal UEFI shell of EDK2 firmware images by locating it on the firmware volume.
+         * As the firmware volume GUID differs by architecture we simply iterate over all volumes
+         * and check if it contains the UEFI shell file.
+         *
+         * On x86 the full device path for the internal UEFI shell is:
+         * Fv(7CB8BDC9-F8EB-4F34-AAEA-3EE4AF6516A1)/FvFile(7C04A583-9E3E-4F1C-AD65-E05268D0B4D1)
+         *
+         * Once located, it is merely a matter to create the device path and pass it to LoadImage().
+         */
+
+        assert(config);
+
+        if (!config->auto_entries || !config->auto_firmware)
+                return;
```

Comment:
i figure this should get its own field, no? i.e. firmware likely has a password set and people know about that it might be accessible. but the uefi shell is normally not expected to be accessible, no? should really have a knob of its own I am sure.

Reply:
i still think this should get its own config file option